cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetSpecificationUpgrader.java

```diff
@@ -98,6 +98,12 @@ public void upgrade() throws Exception {
               for (Map.Entry<Long, byte[]> columnEntry : columnMap.getValue().entrySet()) {
                 Long timeStamp = columnEntry.getKey();
                 byte[] colVal = columnEntry.getValue();
+                // a deleted dataset can still show up here since BufferingTable doesn't actually delete, but
+                // writes a null value. The fact that we need to know that implementation detail here is bad.
+                // If we could use Table here instead of HTable, this would be hidden from us.
+                if (colVal == null || colVal.length == 0) {
```

Comment:
I see; ok. Then its probably fine. @rohitsinha54 just has to keep in mind to do something similar in this PR when he works on his upgrade task.